pkg/cloudprovider/providers/azure/azure_util.go

```diff
@@ -116,6 +125,197 @@ func (az *Cloud) getSecurityRuleID(securityRuleName string) string {
+		lb := (*existingLBs)[lbx]
+		mapExistingLBs[*lb.Name] = &lb
+	}
+	selectedLBRuleCount := math.MaxInt32
+	for asx := range *availabilitySetNames {
+		currASName := (*availabilitySetNames)[asx]
+		currLBName := az.getLoadBalancerName(clusterName, currASName, isInternal)
+		lb, ok := mapExistingLBs[currLBName]
+		if !ok {
+			// select this LB as this is a new LB and will have minimum rules
+			// create tmp lb struct to hold metadata for the new load-balancer
+			selectedLB = &network.LoadBalancer{
+				Name:                         &currLBName,
+				Location:                     &az.Location,
+				LoadBalancerPropertiesFormat: &network.LoadBalancerPropertiesFormat{},
+			}
+
+			return selectedLB, false, nil
+		}
+
+		lbRules := *lb.LoadBalancingRules
+		currLBRuleCount := 0
+		if lbRules != nil {
+			currLBRuleCount = len(lbRules)
+		}
+		if currLBRuleCount < selectedLBRuleCount {
+			selectedLBRuleCount = currLBRuleCount
+			selectedLB = lb
+		}
+	}
+
+	if selectedLB == nil {
+		glog.Errorf("selectLoadBalancer service (%s) - unable to find load balancer for selected availability sets %v", serviceName, *availabilitySetNames)
+		return nil, false, fmt.Errorf("selectLoadBalancer (%s)- unable to find load balancer for selected availability sets %v", serviceName, *availabilitySetNames)
+	}
+	// validate if the selected LB has not exceeded the MaximumLoadBalancerRuleCount
+	if az.Config.MaximumLoadBalancerRuleCount != 0 && selectedLBRuleCount >= az.Config.MaximumLoadBalancerRuleCount {
+		err = fmt.Errorf("selectLoadBalancer service (%s) -  all available load balancers have exceeded maximum rule limit %d", serviceName, selectedLBRuleCount)
+		glog.Error(err)
+		return selectedLB, existsLb, err
+	}
+
+	return selectedLB, existsLb, nil
+}
+
+// getLoadBalancerAvailabilitySetNames selects all possible availability sets for
+// service load balancer, if the service has no loadbalancer mode annotaion returns the
+// primary availability set if service annotation for loadbalancer availability set
+// exists then return the eligible a availability set
+func (az *Cloud) getLoadBalancerAvailabilitySetNames(service *v1.Service, nodes []*v1.Node) (availabilitySetNames *[]string, err error) {
+	hasMode, isAuto, serviceASL := getServiceLoadBalancerMode(service)
+	if !hasMode {
+		// legacy load balancer auto mode load balancer.
+		availabilitySetNames = &[]string{az.Config.PrimaryAvailabilitySetName}
+		return availabilitySetNames, nil
+	}
+	availabilitySetNames, err = az.getAgentPoolAvailabiliySets(nodes)
+	if err != nil {
+		return nil, err
+	}
+	if len(*availabilitySetNames) == 0 {
+		return nil, fmt.Errorf("No availability sets found for nodes, node count(%d)", len(nodes))
+	}
+	// sort the list to have deterministic selection
+	sort.Strings(*availabilitySetNames)
+	if !isAuto {
+		if serviceASL == nil || len(serviceASL) == 0 {
+			return nil, fmt.Errorf("service annotation for LoadBalancerMode is empty, it should have __auto__ or availability sets value")
```

Comment:
if serviceASL validation should be right after if!hasMode..?
__auto__ should be replaced with the value of constant declared somewhere else.

Reply:
In order to validate I need to get the availability set names from the resource backend, adding more comments too